---
title: "OpenArray qPCR Report"
date: "`r Sys.Date()`"
output: html_document
params:
    data_path: NULL
    model_results: NULL
---

```{r include = FALSE}
knitr::opts_chunk$set(echo=FALSE)
```

```{r}
# load in data
se <- readRDS(params$data_path)
pcr_data <- as.data.frame(SummarizedExperiment::colData(se))
```

## Narrative

This report document consolidates run information and QC plots generated from multiplex qPCR data following OpenArray gene expression runs. This document supplements, but does not replace, the tabulated analyzed data output by the core pipeline.

## Graphics

### Fluroescence Curve Quality Control

This graphic plots each qPCR reaction against the 'Crt', 'Amp Score', and 'Cq Conf' output by QuantStudio 12K Flex software. The Crt is the cycle value where fluorescence crosses some threshold value determined from the shape of the amplification curve. Amplification score and Cq confidence represent the quality of the amplification curve and the reliability of the cycle value, respectively. 

For ideal sensitivity and specificity, an amplification score of 1.2 or greater and a Cq confidence of 0.7 or greater is often expected, though the particular thresholds may vary by assay. If a significant proportion of reactions fail to clear these metrics, and the assay is known to generally perform well, an issue with the run should be suspected. Results with a Crt of 28 or higher may represent false positives, and are best interpreted by manual review or sigmoidal curve-fitting. 

```{r, fig.align='center'}
OAtools::plotQC(se)
```

### Amplification Overview

Here we summarize the amplification status of each sample per assay in heat map format. Technical replicates are grouped into single cells where the highest amplification status is considered. i.e. if one replicate is positive and the other negative, the sample is displayed as positive for that assay. 

It is best to ensure here that control material is positive for all expected assays and negative for all other assays, respectively. Control material positive for unexpected assays may indicate off-target amplification effects or potential cross-contamination.

```{r}
OAtools::plotOverview(se)
```

### Crt Distribution

Here is displayed the distribution of observed Crt values by assay target. In general, an approximately normal distribution of Crt values is expected for each assay target over large numbers of samples. Extreme outliers may indicate the need for manual review.

```{r}
OAtools::plotCrt(se)
```

## Results
```{r}

result_table <- pcr_data |> dplyr::group_by(sample_name, target_name)

if (model_results) {
    result_table <- result_table |> dplyr::summarise(
        mean_crt = mean(crt),
        high_amp_status = dplyr::case_when(
            any(.data$amp_status == "Amp") ~ "Amp",
            any(.data$amp_status == "Inconclusive") ~ "Inconclusive",
            TRUE ~ "No Amp"
        ),
        result = dplyr::case_when(
            any(.data$result == "positive") ~ "Positive",
            any(.data$result == "inconclusive") ~ "Inconclusive",
            TRUE ~ "Negative"
        ),
        .groups = "drop"
    )
} else {
    result_table <- result_table |> dplyr::summarise(
        mean_crt = mean(crt),
        high_amp_status = dplyr::case_when(
            any(.data$amp_status == "Amp") ~ "Amp",
            any(.data$amp_status == "Inconclusive") ~ "Inconclusive",
            TRUE ~ "No Amp"
        ),
        .groups = "drop"
    )
}

result_table |> 
    dplyr::mutate(mean_crt = round(mean_crt, 1)) |> 
    knitr::kable(
        format = "html",
        align = rep("l", ncol(result_table))
    ) |> 
    kableExtra::kable_styling(
        bootstrap_options = c("striped", "hover", "bordered"),
        full_width = TRUE
    )
```
